sunpy/io/_asdf.py

```diff
@@ -0,0 +1,116 @@
+from pathlib import Path
+from collections import OrderedDict
+
+import numpy as np
+
+import asdf
+
+from sunpy.io._header import FileHeader
+
+__all__ = ["write","read","get_header","get_keys_name"]
+
+def write(fname, data, header, **kwargs):
+    """
+    Take data and header pairs and save it to asdf file.
+    inspired from https://docs.sunpy.org/en/stable/generated/gallery/saving_and_loading_data/genericmap_in_asdf.html
+
+    format:
+    map_name.asdf (main tree)
+        meta: (header info)
+
+        data : data blocks
+    Parameters
+    ----------
+    fname : `str`
+        File name, with extension.
+    data : `numpy.ndarray`
+        n-dimensional data array.
+    header : `dict`
+        A header dictionary.
+    """
+    map_name = Path(fname)
+    map_name = map_name.name
+    data = data.astype("float64")
+    meta = dict(header)
+    with asdf.AsdfFile() as af:
+        af.tree={map_name:{"meta":meta,"data":data}}
+        af.write_to(fname,**kwargs)
+
+
+def read(fname,**kwargs):
+    """
+    Read a asdf file.
+
+    Parameters
+    ----------
+    filepath : `str`
+        The asdf file to be read.
+
+
+    Returns
+    -------
+    `list`
+        A list of (data, header) tuples
+    """
+
+    with asdf.open(fname) as af:
+        map_name = get_keys_name(fname)
+        try:
```

Comment:
Why is there a try block here?

Reply:
The format in which _asdf/write function saves is 
map_name.asdf (main tree)
                meta (sub tree)
               

                data (sub tree)

there might cases (like in test file for asdf or  https://docs.sunpy.org/en/stable/generated/gallery/saving_and_loading_data/genericmap_in_asdf.html) where the map was stored explicitly in asdf file.

try : this  part handles the asdf file which was saved via _asdf/write method

except : this part where tree was explicitly stored in asdf tree like {object : map}

Reply:
There must be a more elegant way to handle this @Cadair